client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java

```diff
@@ -300,15 +305,43 @@ public <T extends ChannelOperation> T fetchChannelAndInvoke(Set<SocketAddress> f
       } finally {
          lock.writeLock().unlock();
       }
+
+      return recheck
+            ? fetchChannelAndInvoke(server, cacheName, operation)
+            : fetchChannelAndInvoke(server, operation);
+   }
+
+   public <T extends ChannelOperation> T fetchChannelAndInvoke(SocketAddress server, byte[] cacheName, T operation) {
+      boolean suspected;
+      lock.readLock().lock();
+      try {
+         suspected = failedServers.contains(server);
+      } finally {
+         lock.readLock().unlock();
+      }
+
+      if (suspected) {
+         // The next method will acquire the lock when operating in the failedServers.
+         return fetchChannelAndInvoke(failedServers, cacheName, operation, false);
+      }
       return fetchChannelAndInvoke(server, operation);
    }
 
-   public <T extends ChannelOperation> T fetchChannelAndInvoke(SocketAddress server, T operation) {
+   private <T extends ChannelOperation> T fetchChannelAndInvoke(SocketAddress server, T operation) {
       ChannelPool pool = channelPoolMap.computeIfAbsent(server, newPool);
       pool.acquire(operation);
       return operation;
    }
 
+   public void operationFailedForServers(Set<SocketAddress> addresses) {
+      lock.writeLock().lock();
+      try {
+         failedServers.addAll(addresses);
```

Comment:
I don't think we can do this. The `failedServers` collection here is explicitly for servers that received a connection failed event when trying to create a new connection.

The place that we should be seeing if a connection has failed is in the ChannelRecord.operationComplete which is a registered callback when the channel is closed. This in turn calls `channelPool.releaseClosedChannel`